ArduPlane/compassmot.cpp

```diff
@@ -0,0 +1,248 @@
+        compass_mot = false;
+        return MAV_RESULT_TEMPORARILY_REJECTED;
+        }
+    }
+
+    // check throttle is at zero
+    read_radio();
+    if (channel_throttle->get_control_in() != 0) {
+        gcs_chan.send_text(MAV_SEVERITY_CRITICAL, "Throttle not zero");
+        compass_mot = false;
+        return MAV_RESULT_TEMPORARILY_REJECTED;
+    }
+
+    // check we are landed
+    if (is_flying()) {
+        gcs_chan.send_text(MAV_SEVERITY_CRITICAL, "Is flying");
+        compass_mot = false;
+        return MAV_RESULT_TEMPORARILY_REJECTED;
+    }
+
+    // prevent failsafe from triggering due to long delay in main loop
+    // may need to add failsafe_disable() similar to copter
+    in_calibration = true;
+
+    float current;
+
+    // default compensation type to use current if possible
+    if (battery.current_amps(current)) {
+        comp_type = AP_COMPASS_MOT_COMP_CURRENT;
+    } else {
+        comp_type = AP_COMPASS_MOT_COMP_THROTTLE;
+    }
+
+    // send back initial ACK
+    mavlink_msg_command_ack_send(gcs_chan.get_chan(), MAV_CMD_PREFLIGHT_CALIBRATION,0);
+
+    // flash leds
+    AP_Notify::flags.esc_calibration = true;
+
+    // warn user we are starting calibration
+    gcs_chan.send_text(MAV_SEVERITY_INFO, "Starting Calibration");
+
+    // inform what type of compensation we are attempting
+    if (comp_type == AP_COMPASS_MOT_COMP_CURRENT) {
+        gcs_chan.send_text(MAV_SEVERITY_INFO, "Current");
+    } else {
+        gcs_chan.send_text(MAV_SEVERITY_INFO, "Throttle");
+    }
+
+    // disable throttle failsafe
+    g.throttle_fs_enabled = 0;
+  
+    // disable motor compensation
+    compass.motor_compensation_type(AP_COMPASS_MOT_COMP_DISABLED);
+    for (uint8_t i=0; i<compass.get_count(); i++) {
+        compass.set_motor_compensation(i, Vector3f(0,0,0));
+    }
+
+    // get initial compass readings
+    compass.read();
+
+    // store initial x,y,z compass values
+    // initialise interference percentage
+    for (uint8_t i=0; i<compass.get_count(); i++) {
+        compass_base[i] = compass.get_field(i);
+        interference_pct[i] = 0.0f;
+    }
+
+    EXPECT_DELAY_MS(5000);
+
+    // enable motors and pass through throttle
+    init_rc_out_main();
+    hal.util->set_soft_armed(true);
+
+    // initialise run time
+    last_run_time = millis();
+    last_send_time = millis();
+
+    // main run while there is no user input and the compass is healthy
+    while (command_ack_start == command_ack_counter && compass.healthy()) {
+        EXPECT_DELAY_MS(5000);
+
+        // 50hz loop
+        if (millis() - last_run_time < 20) {
+            hal.scheduler->delay(5);
+            continue;
+        }
+        last_run_time = millis();
+
+        // read the radio input
+        read_radio();
+
+        // pass through throttle to motors
+        SRV_Channels::cork();
+        SRV_Channels::set_output_scaled(SRV_Channel::k_throttle, get_throttle_input(true));
```

Comment:
Should also output left and right throttles, we probably also want a option to do CMOT based on forward or vertical motors.